### PR TITLE
Updates and main interface rename

### PR DESCRIPTION
**Pull request by shaynevanasperen**
This pull request should resolve issue #6. I decided to rename `IDispatcher` to `IMagneto` and `IInvoker` to `IMediary`. I decided not to merge those two interfaces after all, but made the implementation of `IMagneto` not require passing an instance of `IMediary` to the constructor so that service registration can be simpler. If consumers of this library do register `IMediary` in the service collection, it will use that, otherwise it creates a new one itself. Since most people will just want to use the main `IMagneto` interface, the "memorability" and "awareness" that @JosephWoodward mentioned should be satisfied. I've kept the method names the same as before: `Query[Async]` and `Command[Async]` because I think it's more expressive that way.

In the process of doing so, I've also made the following changes:
- Updated several NuGet package dependencies, including a bug-fix release of `Code.Extensions.Object.Flatten`
- Added `netstandard2.0` target framework and removed older target frameworks
- Fixed a minor bug (checking of wrong argument in `MemoryCacheStore`)
- Throw more specific exception in `CacheConfig.KeyPrefix` setter when invalid argument is passed
- Remove functionality for optionally caching nulls (always treat nulls as normal query results) and use `IDecorator` for all operation types
- Use _Directory.Build.props_ and version packages independently
- Switched over to the new _Microsoft SourceLink_
- Updated the samples project to _ASP.NET Core 2.1_ and made the examples all use `IMagneto` rather than have some of them use `IMediary`
- Added more XML documentation comments
- Improved cached query state management for cache entry options
- Added `CancellationToken` parameters
- Updated `README` main description and example code

**Due to the vast amount of changes in this pull request, I would advise reviewing each commit separately.**

### CODE REVIEW

File: src/Magneto/Magneto.csproj [hunk 1]
```diff
@@ -2,8 +2,8 @@
 	<Import Project="..\..\Magneto.Common.props" />
 
 	<PropertyGroup>
-		<TargetFrameworks>netstandard1.1;netstandard1.3;net45;net46</TargetFrameworks>
-		<DisableImplicitFrameworkReferences Condition="'$(TargetFramework)' == 'net45' OR '$(TargetFramework)' == 'net46'">true</DisableImplicitFrameworkReferences>
+		<TargetFrameworks>netstandard2.0;net46</TargetFrameworks>
```
Comment (slang25) [inline src/Magneto/Magneto.csproj]: I'm not sure if it's worth the `net46` target. Today I see a lot of people are going full `netstandard2.0`, which will work on 4.6.1 and up.

**Comment by slang25** [comment]
Looking very nice @shaynevanasperen 👍. The new names look sensible too, although I struggle to get excited by names.

File: src/Magneto/Magneto.csproj (hunk 1 above)
Comment (shaynevanasperen) [reply src/Magneto/Magneto.csproj]: Thanks @slang25. I've now removed the `net46` target too, so we're targeting only `netstandard2.0`.

Comment (shaynevanasperen) [reply src/Magneto/Magneto.csproj]: @slang25 Scratch that. Added `net461` back again, according to your own [blog post](https://stuartlang.uk/dotnet-framework-support-for-dotnetnet-standard-2-0/) about it.

**Comment by shaynevanasperen** [comment]
@slang25 @JosephWoodward Please review this for me.

File: src/Magneto/Core/CacheEntry.cs [hunk 2]
```diff
@@ -4,4 +4,12 @@ public class CacheEntry<T>
 	{
 		public T Value { get; set; }
 	}
+
+	public static class CacheEntryExtensions
```
Comment (slang25) [inline src/Magneto/Core/CacheEntry.cs]: Pro tip (from Martin C), add `[EditorBrowsable(EditorBrowsableState.Never)]` to classes in libraries that contain only extension methods. Keeps the browsable apis down in intellisense.

File: src/Magneto/CachedQuery.cs [hunk 3]
```diff
@@ -66,8 +67,8 @@ public abstract class AsyncCachedQuery<TContext, TCacheEntryOptions, TResult> :
 		Core.AsyncCachedQuery<TContext, TCacheEntryOptions, TResult>, IAsyncCachedQuery<TContext, TCacheEntryOptions, TResult>
 	{
 		/// <inheritdoc cref="IAsyncCachedQuery{TContext,TCacheEntryOptions,TResult}.ExecuteAsync"/>
-		public virtual Task<TResult> ExecuteAsync(TContext context, IAsyncCacheStore<TCacheEntryOptions> cacheStore, CacheOption cacheOption = CacheOption.Default) =>
-			GetCachedResultAsync(context, cacheStore, cacheOption);
+		public virtual Task<TResult> ExecuteAsync(TContext context, IAsyncCacheStore<TCacheEntryOptions> cacheStore, CancellationToken cancellationToken = default, CacheOption cacheOption = CacheOption.Default) =>
```
Comment (slang25) [inline src/Magneto/CachedQuery.cs]: It's common practice to have `cancellationToken` as the last parameter, is there a good reason not to follow that here?

File: src/Magneto/Core/CacheConfig.cs [hunk 4]
```diff
@@ -43,7 +40,6 @@ public virtual object VaryBy
 		}
 		object _varyBy;
 
-		/// <inheritdoc />
 		public virtual string Key => _key ?? (_key = _createKey(KeyPrefix, VaryBy));
 
 		string buildKey(string keyPrefix, object varyBy)
```
Comment (slang25) [inline src/Magneto/Core/CacheConfig.cs]: Not part of this PR, but get rid of the new List in this method, and do:
```csharp
return KeyPrefix + string.Join("_", VaryBy.Flatten());
```
for less allocations.

File: src/Magneto/Core/CacheEntry.cs (hunk 2 above)
Comment (shaynevanasperen) [reply src/Magneto/Core/CacheEntry.cs]: Thanks. Done!

File: src/Magneto/CachedQuery.cs (hunk 3 above)
Comment (shaynevanasperen) [reply src/Magneto/CachedQuery.cs]: Thanks. I did read that article, and I'd decided to not put the `cancellationToken` parameter last, because at the time I thought that it would be more likely that consumers would want to only pass the `cancellationToken` and not the other optional argument (`cacheOption`). This would have meant that they could avoid naming the argument at the call site. However, upon reading your comment and thinking about it some more, I now think the opposite is true (they'd more likely want to pass `cacheOption` and not `cancellationToken`).

File: src/Magneto/Core/CacheConfig.cs (hunk 4 above)
Comment (shaynevanasperen) [reply src/Magneto/Core/CacheConfig.cs]: Thanks. I've incorporated this suggested change.

**Comment by JosephWoodward** [comment]
A little late but looks good to me, the `IMagneto` name change is loads better. Going to go and update my projects that use Magneto now.
